Bump to Go 1.19, Ginkgo v2 & kubebuilder 3.7.1

laxmikantbpandhare opened the pull request:
**Description of the change:**
Changes in this PR are:

- Modified go version to 1.19 in all the places and Kubebuilder to k8s 1.25
- The `config-gen` is completely removed from Kubebuilder. As part of this bump, the website content of SDK was updated where the doc is referring to `config-gen`. 
- Updated ginkgo to ginkgo/v2 in all of the files. The reason behind it is that Kubebuilder updated to ginkgo/v2.
- Updated all the `io/ioutil` imported packages to either `io` or `os`. `io/ioutil` got deprecated in 1.16 and no longer it is used. Updated all the files in SDK.
- Ansible scaffolding got modified in order to incorporate with Kubebuilder `--component-config` flag changes. Previously, it was scaffolding and doing some file changes without considering the `--component-config` flag.
- This PR bumped Kubebuilder to a newer version and it got the recent PR changes for the `--component-config` flag.
- Helm scaffolding got modified in order to incorporate with Kubebuilder `--component-config` flag changes. Previously, it was scaffolding and doing some file changes without considering the `--component-config` flag.



**Motivation for the change:**

- Ginkgo version mismatch in Kubebuilder and `operator-sdk`. `ginkgo` updated to `ginkgo/v2`
- Updated `io/ioutil` to `io` or `oc` as `io/ioutil` got deprecated in `1.16`
- Updated go version to `1.19`

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [x] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

laxmikantbpandhare commented:
Updated the `Kubebuilder` version as the new KB version has `ginkgo/v2` changes.

laxmikantbpandhare commented:
/hold

laxmikantbpandhare commented:
This one is blocked by KB PR https://github.com/kubernetes-sigs/kubebuilder/pull/3010

varshaprasad96 commented on testdata/ansible/memcached-operator/config/default/kustomization.yaml:
```diff
@@ -18,6 +18,11 @@ resources:
 - ../crd
 - ../rbac
 - ../manager
+# [WEBHOOK] To enable webhook, uncomment all the sections with [WEBHOOK] prefix including the one in
```
Are we starting to support webhooks and cert-manager by default in ansible and helm operators? If so, this again needs a migration guide entry.

laxmikantbpandhare replied:
Let me check on this.

laxmikantbpandhare replied:
same like this - https://github.com/operator-framework/operator-sdk/pull/6047#discussion_r993834871

laxmikantbpandhare replied:
Updated

laxmikantbpandhare commented on internal/bindata/olm/manifests.go:
```diff
@@ -316,7 +315,7 @@ func RestoreAsset(dir, name string) error {
 		return err
 	}
-	err = ioutil.WriteFile(_filePath(dir, name), data, info.Mode())
+	err = os.WriteFile(_filePath(dir, name), data, info.Mode())
```
agree, I modified all the files. I will remove this change.

laxmikantbpandhare replied:
removed this.

laxmikantbpandhare commented on hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go:
```diff
@@ -71,15 +71,17 @@ func (mh *Memcached) Run() {
 			"--repo", "github.com/example/memcached-operator",
-			"--domain", mh.ctx.Domain)
+			"--domain", mh.ctx.Domain,
+			"--component-config=true")
```
Yes, without it `component-manager-config` file will not get scaffolded as it was previously by default used to scaffold. If we don't add this then test cases will fail.

laxmikantbpandhare replied:
removed this as we modified coding functionality.

laxmikantbpandhare commented on testdata/helm/memcached-operator/config/default/kustomization.yaml:
```diff
@@ -18,6 +18,11 @@ resources:
 - ../crd
 - ../rbac
 - ../manager
+# [WEBHOOK] To enable webhook, uncomment all the sections with [WEBHOOK] prefix including the one in
```
this is one of the changes I added and it is causing this `webhook` and `certmanager` scaffolding. I will check it only for `--leader-elect`.

```
if !p.config.IsComponentConfig() {
		if err := addInitCustomizations(p.config.GetProjectName()); err != nil {
			return fmt.Errorf("error updating init manifests: %s", err)
		}
	}
```

laxmikantbpandhare replied:
Updated.

laxmikantbpandhare commented on testdata/go/v3/memcached-operator/main.go:
```diff
@@ -69,25 +66,17 @@ func main() {
-		// after the manager stops then its usage might be unsafe.
-		// LeaderElectionReleaseOnCancel: true,
-	})
+	var err error
```
these `main.go` changes gets added after inclusion of `--component-config` flag in Kubebuilder.

theishshah commented:
/lgtm

laxmikantbpandhare commented on testdata/go/v3/memcached-operator/Makefile:
```diff
@@ -126,6 +129,23 @@ docker-build: test ## Build docker image with the manager.
+# - have enable BuildKit, More info: https://docs.docker.com/develop/develop-images/build_enhancements/
+# - be able to push the image for your registry (i.e. if you do not inform a valid value via IMG=<myregistry/image:<tag>> than the export will fail)
+# To properly provided solutions that supports more than one platform you should use this option.
+PLATFORMS ?= linux/arm64,linux/amd64,linux/s390x,linux/ppc64le
```
Can we add this as a new feature request? I don't want to add any more changes to this PR. This PR already becomes huge in terms of changes. @jmrodri WDYT?

laxmikantbpandhare replied:
sure.

everettraven commented on changelog/fragments/ginkgo-v2-changes.yaml:
```diff
@@ -0,0 +1,16 @@
+    kind: "change"
+
+    # Is this a breaking change?
+    breaking: false
```
Since there are a lot more changes going in here that affect users we need to add more changelog entries that include the changes made and any associated migration guides.

laxmikantbpandhare replied:
Yes, will add separate changelogs with migration guides.

everettraven commented on internal/plugins/ansible/v1/init.go:
```diff
@@ -243,5 +265,13 @@ func addInitCustomizations(projectName string) error {
 		return fmt.Errorf("error updating kustomization.yaml files: %v", err)
 	}
 
+	const target = "      # according to the platforms which are supported by your solution. "
+	const replace = "      # according to the platforms which are supported by your solution."
+
+	err = util.ReplaceInFile(managerFile, target, replace)
+	if err != nil {
+		return err
+	}
```
Can we add a TODO comment here to remove this whenever the scaffolding is updated in Kubebuilder to have proper yaml output (by proper I mean doesn't have lint errors)?

jmrodri replied:
Also let's create an issue against kubebuilder to have that fixed. Reference this PR in the issue.

laxmikantbpandhare replied:
Agree added like below.

`//TODO: Remove below code snippet whenever the scaffolding is updated in Kubebuilder to have proper YAML output.`

everettraven replied:
I will create an issue upstream. I think we should create the issue as a "we need to add a yaml linter" so that we don't run into these types of issues again and we catch any of these problems in kubebuilder before they are included in a release.

laxmikantbpandhare replied:
Up, I will do that

everettraven commented on internal/plugins/ansible/v1/init.go:
```diff
@@ -213,20 +230,25 @@ func addInitCustomizations(projectName string) error {
+		err = util.ReplaceInFile(managerProxyPatchFile, "8081", "6789")
+		if err != nil {
+			return err
+		}
 	}
-	// Remove the webhook option for the componentConfig since webhooks are not supported by ansible
-	err = util.ReplaceInFile(managerConfigFile, "webhook:\n  port: 9443", "")
-	if err != nil {
-		return err
+
+	if componentConfig {
+		managerConfigFile := filepath.Join("config", "manager", "controller_manager_config.yaml")
+		err = util.ReplaceInFile(managerConfigFile, "8081", "6789")
+		if err != nil {
+			return err
+		}
+		// Remove the webhook option for the componentConfig since webhooks are not supported by ansible
+		err = util.ReplaceInFile(managerConfigFile, "webhook:\n  port: 9443", "")
+		if err != nil {
+			return err
+		}
```
Ideally we can condense this into the one if-else block above. I am fine with this going in like this because it preserves the same logic flow just wrapping it with conditionals. Just figured that I would mention that we should ideally come back and clean this up a bit :).

laxmikantbpandhare replied:
updated in the if-else block.

everettraven commented on internal/plugins/helm/v1/init.go:
```diff
@@ -218,5 +234,13 @@ func addInitCustomizations(projectName string) error {
 		return fmt.Errorf("error updating kustomization.yaml files: %v", err)
 	}
 
+	const target = "      # according to the platforms which are supported by your solution. "
+	const replace = "      # according to the platforms which are supported by your solution."
+
+	err = util.ReplaceInFile(managerFile, target, replace)
+	if err != nil {
+		return err
+	}
+
```
Same as the ansible plugin, lets add a TODO comment here to make sure we clean this up when Kubebuilder scaffolds are fixed.

laxmikantbpandhare replied:
Updated.

everettraven commented on testdata/ansible/memcached-operator/bundle/manifests/cache.example.com_memcacheds.yaml:
```diff
@@ -1,50 +0,0 @@
-apiVersion: apiextensions.k8s.io/v1
```
Why are many of the files under the `manifests/` directory getting removed for the ansible testdata?

everettraven commented on testdata/ansible/memcached-operator/bundle/tests/scorecard/config.yaml:
```diff
@@ -1,70 +0,0 @@
-apiVersion: scorecard.operatorframework.io/v1alpha3
```
Why did this file get deleted?

laxmikantbpandhare replied:
Updated test data properly.

everettraven commented on testdata/helm/memcached-operator/bundle/manifests/cache.example.com_memcacheds.yaml:
```diff
@@ -1,50 +0,0 @@
-apiVersion: apiextensions.k8s.io/v1
```
Why are a lot of the files under the `manifests/` directory deleted for the helm testdata?

laxmikantbpandhare replied:
Updated test data properly.

everettraven reviewed (changes requested):
A couple things:

- We need to update the changelog to include entries for everything that is a user facing change and may require a migration guide.
- I would like add comments to the helm/ansible plugins where we are replacing code because of the yaml linting error. Ideally this will be fixed in Kubebuilder at some point and we won't need to do this replace anymore.
- The ansible/helm samples in the `testdata/` directory have many files deleted from the `manifests/` directory. Why is this being done?

There may be a couple other little nits, but the above summary is the major things that I think we need to address before merging this in.

jmrodri commented:
Saw a typo in generation code, it's been there a while so I'm not going to fix it in this PR. https://github.com/operator-framework/operator-sdk/issues/6085

jmrodri commented on hack/generate/samples/internal/go/memcached-with-webhooks/e2e_test_code.go:
```diff
@@ -35,7 +34,7 @@ func (mh *Memcached) implementingE2ETests() {
 func (mh *Memcached) addTestE2eMaekefileTarget() {
 	err := kbutil.ReplaceInFile(filepath.Join(mh.ctx.Dir, "Makefile"),
-		`KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test ./... -coverprofile cover.out`,
+		`KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./... -coverprofile cover.out`,
```
I think this needs to be called out in a migration.

laxmikantbpandhare replied:
Added migration guide for it.

jmrodri commented on testdata/go/v3/memcached-operator/Makefile:
```diff
@@ -49,7 +49,7 @@ endif
 IMG ?= controller:latest
 # ENVTEST_K8S_VERSION refers to the version of kubebuilder assets to be downloaded by envtest binary.
-ENVTEST_K8S_VERSION = 1.24.2
+ENVTEST_K8S_VERSION = 1.25.0
```
Ugh nevermind, this is from kubebuilder

laxmikantbpandhare replied:
Yes, test data is automatically bumped to 1.25 via KB

jmrodri reviewed (changes requested):
Great PR! That was a lot of work. A few things I think we need to add to this (this is a short summary, use the comments in the review for all of the actual details) :) 

* update PR description to give a summary of all of the changes in this PR. You bumped go, ginkgo, and kube. You had to change some Labels to LabelsMap, explain why that had to happen. Mention some of the repercussions the kube bump had with component config.
* I think we need a couple changelog files. Might be easier to capture the migrations for the specific ones if they are in separate files
  * Changes to ansible scaffolding, with migration
  * Changes to helm scaffolding, with migration
  * Changes to go dependencies, with migration; also call out the multiarch change. That might have to be a migration step as well.
  * Component Config needs to be mentioned somewhere, either in the ansible and helm or separately
  * go, ginkgo, kubebuilder bumps can be in a single changelog fragment

laxmikantbpandhare commented on internal/olm/operator/helpers.go:
```diff
@@ -34,7 +34,7 @@ func CatalogNameForPackage(pkg string) string {
 
 // LoadBundle returns metadata and manifests from within bundleImage.
-func LoadBundle(ctx context.Context, bundleImage string, skipTLSVerify bool, useHTTP bool) (registryutil.Labels, *apimanifests.Bundle, error) {
+func LoadBundle(ctx context.Context, bundleImage string, skipTLSVerify bool, useHTTP bool) (registryutil.LabelsMap, *apimanifests.Bundle, error) {
```
Yes, the `ginkgo/v2` new changes have their own Labels field which was causing the issue.

laxmikantbpandhare commented on internal/registry/labels.go:
```diff
@@ -34,22 +34,22 @@ func (e MetadataNotFoundError) Error() string {
 }
 
 // Labels is a set of key:value labels from an operator-registry object.
-type Labels map[string]string
+type LabelsMap map[string]string
```
Sure.

laxmikantbpandhare replied:
Updated.

`This changes raised one issue where Labels field is used by ginkgo/v2. This field is already defined in SDK which was causing the issue. I updated the SDK field Labels to LabelsMap to avoid the issue causing due to same name.
`

laxmikantbpandhare commented on testdata/go/v3/memcached-operator/go.mod:
```diff
@@ -1,22 +1,22 @@
+	k8s.io/api v0.25.0
+	k8s.io/apimachinery v0.25.0
+	k8s.io/client-go v0.25.0
+	sigs.k8s.io/controller-runtime v0.13.0
```
will add it in migration guide.

laxmikantbpandhare commented on internal/plugins/ansible/v1/init.go:
```diff
@@ -151,25 +152,41 @@ func (p *initSubcommand) PostScaffold() error {
-		fmt.Sprintf("\n        - \"--leader-election-id=%s\"", projectName))
-	if err != nil {
-		return err
+	if componentConfig {
+		err := util.InsertCode(managerFile,
+			"- /manager",
+			fmt.Sprintf("\n        args:\n        - --leader-election-id=%s", projectName))
+		if err != nil {
+			return err
+		}
+
+		err = util.InsertCode(managerProxyPatchFile,
+			"memory: 64Mi",
+			fmt.Sprintf("\n      - name: manager\n        args:\n        - \"--leader-election-id=%s\"", projectName))
+		if err != nil {
+			return err
+		}
+	} else {
+		err := util.InsertCode(managerFile,
+			"--leader-elect",
+			fmt.Sprintf("\n        - --leader-election-id=%s", projectName))
+		if err != nil {
+			return err
+		}
+		err = util.InsertCode(managerProxyPatchFile,
+			"- \"--leader-elect\"",
+			fmt.Sprintf("\n        - \"--leader-election-id=%s\"", projectName))
+		if err != nil {
+			return err
+		}
```
Yup.

laxmikantbpandhare replied:
added step by step explanation

```
      1. If `--component-config` flag is passed then look for `- /manager` in the file in order to add leader election id into the manager.yaml file. If flag is not passed then look for `--leader-elect`.
      2. Same logic for manager proxy patch file, look for `memory: 64Mi` if flag is passed. If flag is not passed then look for `--leader-elect`
      3. Repeat the same logic for replacing the port number from the file.
      4. At the end, update `manager.yaml` file to avoid lint errors.
```

laxmikantbpandhare commented:
Added changelogs below:

One change log for go bump to 1.19, Kubebuilder bump, and ginkgo/v2 changes and added two migration guides to it

- For multi-arch support
- Makefile scaffolding for `--bin-dir $(LOCALBIN)`

laxmikantbpandhare commented on internal/plugins/helm/v1/init.go:
```diff
@@ -130,7 +130,8 @@ func (p *initSubcommand) InjectConfig(c config.Config) error {
 func (p *initSubcommand) Scaffold(fs machinery.Filesystem) error {
-	if err := addInitCustomizations(p.config.GetProjectName()); err != nil {
+
+	if err := addInitCustomizations(p.config.GetProjectName(), p.config.IsComponentConfig()); err != nil {
```
repeated here as well.

jmrodri commented:
/hold

I'm still reviewing the latest changes.

varshaprasad96 commented on changelog/fragments/ginkgoV2-go-kb-bump-changes.yaml:
````diff
@@ -0,0 +1,50 @@
+            ```sh
+                `KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./... -coverprofile cover.out`,
+            ```
+        2. Kubebuiler bump also added multiple architecture support changes to the Makefile whenre docker buildx is used.  This change will support more than one platform.
````
Can we break this changes. Point (1) `addTestE2eMaekefileTarget` doesn't seem to be user facing, right?
Point (2) are the Makefile changes to support multiple arch in `go/v3` plugin. So let's have the title to:
- Add support for building binary in multiple platforms.
And in the migration guide let's add the `git diff` output, to explicitly mention the additions. This is an important addition, so would be nice to highlight it as a separate item.
The same thing for point (3). Let's have the title format to be:
(go/v3)(xxx plugin) Bump dependencies in `go.mod` file

laxmikantbpandhare replied:
Updated with `git diff`

jmrodri commented on changelog/fragments/ansible-scaffolding-changes.yaml:
```diff
@@ -0,0 +1,24 @@
+# release notes and/or the migration guide
+entries:
+  - description: >
+      Ansible scaffolding got modified in order to incorporate with Kubebuilder `--component-config` flag changes. Previously, it was scaffolding and doing some file changes without considering the `--component-config` flag.
+      Now, this PR made changes that will consider the `--component-config` flag and modify the scaffolded file.
+
+      1. If the `--component-config` flag is passed then look for `- /manager` in the file in order to add leader election id to the manager.yaml file. If flag is not passed then look for `--leader-elect`.
+      2. Same logic for manager proxy patch file, look for `memory: 64Mi` if the flag is passed. If a flag is not passed then look for `--leader-elect`
+      3. Repeat the same logic for replacing the port number from the file.
+      4. At the end, update the `manager.yaml` file to avoid lint errors.
```
I'm not sure people really care that we are looking for `--leader-elect` vs `memory: 64Mi`. What they care about is knowing how this change might affect them. And it largely doesn't. The values will be there just in a different place. So calling out it is a different area and that a flag is causing it, might tell them why something seems to have changed.

```suggestion
(ansible/v1) Ansible plugin now takes the `--component-config` into consideration.
    1. The plugin will add the leader election id to the manager.yaml file in
       a different area of the `- /manager` section depending on the value of the 
       component-config flag.
    2. Same logic for the manager proxy patch file, depending on the value of
       the component-config flag, we update the file in a different area.
    3. Port numbers are replaced in a different area depending on the value of
       the component-config flag.
    4. Because of a bug in Kubebuilder, we patch the `manager.yaml` to avoid
       lint errors.
```

laxmikantbpandhare replied:
This one I updated according to Varsha's comment as you suggested on slack chat.

varshaprasad96 reviewed (changes requested):
@laxmikantbpandhare great work! Thanks for handling this big chunk of updates. Just a few changes in changeling fragments - It would be nice to include a snippet of `git diff` in migration sections, so that its easier for users to know what is getting added and/or deleted. Rest of the changes in SDK and plugins look good!

laxmikantbpandhare commented on internal/bindata/olm/manifests.go:
```diff
@@ -255,11 +255,13 @@ var _bindata = map[string]func() (*asset, error){
-//       img/
-//         a.png
-//         b.png
+//
```
Yes.

jmrodri commented on changelog/fragments/ansible-scaffolding-changes.yaml:
````diff
@@ -0,0 +1,24 @@
+    kind: "change"
+    breaking: false
+    migration:
+      header: this change also modified ansible molecule test that was looking for `--leader-elect` in the file.
+      body: |
+        In the advance molecule test, modified code to behave properly with new `--component-config` flag changes.
+
+        ```
+          	err = kbutil.InsertCode(
+                    filepath.Join(ma.ctx.Dir, "config", "default", "manager_auth_proxy_patch.yaml"),
+                    "- \"--leader-election-id=advanced-molecule-operator\"",
+                    managerAuthArgs)
````
Isn't the advanced molecule test part of our internal hack script? This does not have any bearing on the users ansible project.

The purpose of the migration is to signify to the user what they would have to change in their project to match a newly scaffolded project.

For example, let's say in previous version we output two files. A foobar.yaml and a baz.yaml.

Now let's say in this PR, we removed `baz.yaml` and we changed the contents of the `foobar.yaml` by adding the following line `- /manager`. 

So what you would say in the migration step is:
1. Remove the baz.yaml it is no longer needed with this version of the SDK
2. After section SOMETHING, add `- /manager`.

The reason we tell them to do this is so that in the future if there were changes to `create api` or other plugins that look at the contents of those files they can see files that look as similar to what the new scaffolds look like.

laxmikantbpandhare replied:
removed this code snippet as it is part of testdata generation.

```
    migration:
      header: this change also modified ansible molecule test that was looking for `--leader-elect` in the file.
      body: |
        In the advance molecule test, modify the below code to behave properly with new `--component-config` flag changes.
        
        ```go
          	"- \"--leader-election-id=advanced-molecule-operator\"",
        ``` 
```

laxmikantbpandhare replied:
got it. thank you for a detailed explanation. :)

jmrodri commented on changelog/fragments/component-config-changes.yaml:
```diff
@@ -0,0 +1,10 @@
+# release notes and/or the migration guide
+entries:
+  - description: >
+      This PR bumped Kubebuilder to a newer version and it got the recent PR changes for the `--component-config` flag.
```
```suggestion
      Bumped Kubebuilder to 3.7.1, with this version comes the addition of the `--component-config` flag. 
```

jmrodri commented on changelog/fragments/component-config-changes.yaml:
```diff
@@ -0,0 +1,10 @@
+  - description: >
+      This PR bumped Kubebuilder to a newer version and it got the recent PR changes for the `--component-config` flag.
+
+      If this flag is passed through the command line as an argument then only the `controller_manager_config.yaml` file will be scaffolded in the manifest. 
```
If the flag is passed to WHAT? Under what conditions is the flag used? What commands can I pass it in?

laxmikantbpandhare replied:
Updated the content as below.

```
      This PR bumped Kubebuilder to 3.7.1,  with this version comes the addition of the `--component-config` flag. 

      This flag is passed through `operator-sdk init` command. 
      If this flag is passed through the command line as an argument then only the 
      `controller_manager_config.yaml` file will be scaffolded in the 
      manifest. 
      If it is not passed then it will not scaffold out this file. 
      But, in this case, it will add `--leader-elect` to the `manager.yaml`.  
      The command for passing the flag is: 
      operator-sdk init --domain example.com --repo github.com/example/memcached-operator --component-config  
```

jmrodri commented on changelog/fragments/component-config-changes.yaml:
```diff
@@ -0,0 +1,10 @@
+      This PR bumped Kubebuilder to a newer version and it got the recent PR changes for the `--component-config` flag.
+
+      If this flag is passed through the command line as an argument then only the `controller_manager_config.yaml` file will be scaffolded in the manifest. 
+      If it is not passed then it will not scaffold out this file. But, in this case, it will add `--leader-elect` to the files to the `manager.yaml` file
```
What adds `--leader-elect`? All the plugins? some? Is that relevant to this portion?

laxmikantbpandhare replied:
removed.

jmrodri commented on changelog/fragments/ginkgoV2-go-kb-bump-changes.yaml:
```diff
@@ -0,0 +1,50 @@
+# release notes and/or the migration guide
+entries:
+  - description: >
+      Modified go version to 1.19 in all the places and Kubebuilder to k8s 1.25
```
Did we bump to k8s 1.25? because the go.mod still says 0.24.

jmrodri replied:
If this is the ginkgo, kb and go bump, why isn't ginkgo mentioned here?

laxmikantbpandhare replied:
updated as Modified go version to 1.19 and Kubebuilder to 3.7.1

Created 3 different entries.

I wanted to add a migration guide separately to kb bump 3.7.1

jmrodri commented on changelog/fragments/ginkgoV2-go-kb-bump-changes.yaml:
```diff
@@ -0,0 +1,50 @@
+entries:
+  - description: >
+      Modified go version to 1.19 in all the places and Kubebuilder to k8s 1.25
+      The `config-gen` is completely removed from Kubebuilder. As part of this bump, the website content of SDK was updated where the doc is referring to `config-gen`. 
```
This should be part of the kubebuilder bump changelog you had above.

jmrodri commented on changelog/fragments/ginkgoV2-go-kb-bump-changes.yaml:
````diff
@@ -0,0 +1,50 @@
+    migration:
+      header: Kubebuilder bump brought changes that modified the `Makefile` scaffolding and also `multiarch` chanegs are also included.
+      body: | 
+        1. Updated the addTestE2eMaekefileTarget() target where Makefile content is used as below.
+
+            ```sh
+                `KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./... -coverprofile cover.out`,
+            ```
````
`addTestE2eMaekefileTarget` is an internal hack method. Users don't care about these changes.

jmrodri replied:
what they do care about is when you scaffold out a project does the Makefile contain the change above? I.e. if I scaffold out a project with operator-sdk 1.23 and one containing this patch, will the Makefiles look different? If so, that is what you need to include in the migration steps.

laxmikantbpandhare replied:
removed `addTestE2eMaekefileTarget ` it completely and updated the content.

jmrodri commented on changelog/fragments/ginkgoV2-go-kb-bump-changes.yaml:
````diff
@@ -0,0 +1,50 @@
+            ```sh
+                `KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./... -coverprofile cover.out`,
+            ```
+        2. Kubebuiler bump also added multiple architecture support changes to the Makefile whenre docker buildx is used.  This change will support more than one platform.
````
typo: whenre should be where

laxmikantbpandhare replied:
updated according to varsha comment

jmrodri commented on changelog/fragments/ginkgoV2-go-kb-bump-changes.yaml:
````diff
@@ -0,0 +1,50 @@
+            ```
+        2. Kubebuiler bump also added multiple architecture support changes to the Makefile whenre docker buildx is used.  This change will support more than one platform.
+
+            ```sh
+            PLATFORMS ?= linux/arm64,linux/amd64,linux/s390x,linux/ppc64le
+            .PHONY: docker-buildx
+            docker-buildx: test ## Build and push docker image for the manager for cross-platform support
+              # copy existing Dockerfile and insert --platform=${BUILDPLATFORM} into Dockerfile.cross, and preserve the original Dockerfile
+              sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/' Dockerfile > Dockerfile.cross
+              - docker buildx create --name project-v3-builder
+              docker buildx use project-v3-builder
+              - docker buildx build --push --platform=$(PLATFORMS) --tag ${IMG} -f Dockerfile.cross
+              - docker buildx rm project-v3-builder
+              rm Dockerfile.cross
+            ```
````
We need to tell them where to make these changes in their project. Is it a new target? If so, then tell them to add it in their Makefile at the end or after an existing section.

If this is an existing target, then tell them how they can modify their target to match the new one we expect.

laxmikantbpandhare replied:
in the Makefile. Added like below

Add below content to the Makefile.

jmrodri commented on changelog/fragments/ginkgoV2-go-kb-bump-changes.yaml:
````diff
@@ -0,0 +1,50 @@
+                sigs.k8s.io/controller-runtime v0.13.0
+            ```
+  - description: >
+      Updated ginkgo to ginkgo/v2 in all of the files. The reason behind it is Kubebuilder updated to ginkgo/v2.
+      These changes raised one issue where the Labels field is used by ginkgo/v2. Same field name is already defined in SDK that was causing the issue. I updated the SDK field Labels to LabelsMap to avoid the issue caused due to the same name.
````
Do external users care about the Labels change? What they care about is how this change will affect them directly. Think in the context of the projects that were scaffolded. How will ginkgo/v2 affect them?

laxmikantbpandhare replied:
Again, it was the suggestion in previous review. I will remove it completely.

jmrodri commented on changelog/fragments/ginkgoV2-go-kb-bump-changes.yaml:
```diff
@@ -0,0 +1,50 @@
+    kind: "change"
+    breaking: false
+  - description: >
+      Updated all the `io/ioutil` imported package to either `io` or `os`. `io/ioutil` got deprecated in 1.16 and no longer it is used. Updated all the files in SDK.
```
```suggestion
      As part of the Go 1.19 update, you may find that calls to `io/ioutil` will no longer work in your project. You will have to migrate them to either `io` or `os` depending on the methods you are using. The `io/ioutil` package was deprected in Go 1.16 and removed in 1.19.
```

laxmikantbpandhare replied:
Updated according to Varsha's comment.

jmrodri commented on changelog/fragments/helm-scaffolding-changes.yaml:
```diff
@@ -0,0 +1,13 @@
+# release notes and/or the migration guide
+entries:
+  - description: >
+      Helm scaffolding got modified in order to incorporate with Kubebuilder `--component-config` flag changes. Previously, it was scaffolding and doing some file changes without considering component-config flag.
+      Now, this PR made changes which will consider the `--component-config` flag and modify the scaffolded file.
+
+      1. If the `--component-config` flag is passed then look for `- /manager` in the file in order to add leader election id to the manager.yaml file. If flag is not passed then look for `--leader-elect`.
+      2. Same logic for manager proxy patch file, look for `memory: 64Mi` if the flag is passed. If a flag is not passed then look for `--leader-elect`
+      3. Repeat the same logic for replacing the port number from the file.
+      4. At the end, update the `manager.yaml` file to avoid lint errors.
```
See my comments about ansible above. Same things apply to helm

laxmikantbpandhare replied:
udpated

jmrodri reviewed (changes requested):
The changelog is meant to convey what things have changed that would affect a user's project. They don't care about changes we made to internal code. 

How will go 1.19 update impact a users project? the io/ioutil is a good example.

How will ginkgo v2 affect them? They dont' care that we had to rename an internal variable like LabelsMap. They care that they might have to change their suite_test.go. So look at the difference of the testdata changes. THAT is usually what they care to see. So if the testdata changes the Makefile, what changed? That is what they will want to see.

Also, migrations are only important *if* it affects the scaffolded out project. If the change has no effect then no need for a migration.

I know it's not easy but you'll want to think about each of these as you write the changelogs. Usually the changelogs are easy because the changes in the PR aren't that big. But this is a HUGE PR and a lot changed :)

jmrodri commented on changelog/fragments/ginkgoV2-go-kb-bump-changes.yaml:
````diff
@@ -0,0 +1,50 @@
+                sigs.k8s.io/controller-runtime v0.13.0
+            ```
+  - description: >
+      Updated ginkgo to ginkgo/v2 in all of the files. The reason behind it is Kubebuilder updated to ginkgo/v2.
+      These changes raised one issue where the Labels field is used by ginkgo/v2. Same field name is already defined in SDK that was causing the issue. I updated the SDK field Labels to LabelsMap to avoid the issue caused due to the same name.
+    kind: "change"
+    breaking: false
+  - description: >
+      Updated all the `io/ioutil` imported package to either `io` or `os`. `io/ioutil` got deprecated in 1.16 and no longer it is used. Updated all the files in SDK.
+    kind: "change"
+    breaking: false
````
@laxmikantbpandhare I like Varsha's comments. Much simpler than even mine.

laxmikantbpandhare replied:
Updated

everettraven commented on changelog/fragments/ginkgoV2-go-kb-bump-changes.yaml:
````diff
@@ -0,0 +1,50 @@
+            ```sh
+                `KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./... -coverprofile cover.out`,
+            ```
+        2. Kubebuiler bump also added multiple architecture support changes to the Makefile whenre docker buildx is used.  This change will support more than one platform.
````
```suggestion
        2. Kubebuilder bump also added multiple architecture support changes to the Makefile where docker buildx is used.  This change will support more than one platform.
```

laxmikantbpandhare replied:
updated

laxmikantbpandhare commented on changelog/fragments/ansible-scaffolding-changes.yaml:
```diff
@@ -0,0 +1,24 @@
+# release notes and/or the migration guide
+entries:
+  - description: >
+      Ansible scaffolding got modified in order to incorporate with Kubebuilder `--component-config` flag changes. Previously, it was scaffolding and doing some file changes without considering the `--component-config` flag.
+      Now, this PR made changes that will consider the `--component-config` flag and modify the scaffolded file.
+
```
Updated.

laxmikantbpandhare commented on changelog/fragments/ansible-scaffolding-changes.yaml:
````diff
@@ -0,0 +1,24 @@
+      body: |
+        In the advance molecule test, modified code to behave properly with new `--component-config` flag changes.
+
+        ```
````
the test case just got modified as the code got modified. Yes, will update the `git diff` output.

laxmikantbpandhare commented on changelog/fragments/ansible-scaffolding-changes.yaml:
```diff
@@ -0,0 +1,24 @@
+    migration:
+      header: this change also modified ansible molecule test that was looking for `--leader-elect` in the file.
+      body: |
+        In the advance molecule test, modified code to behave properly with new `--component-config` flag changes.
```
updated

jmrodri commented:
/hold cancel

I finished my review

everettraven reviewed (approved):
/lgtm
